network/msgCompressor.go

```diff
@@ -0,0 +1,151 @@
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/DataDog/zstd"
+
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+var zstdCompressionMagic = [4]byte{0x28, 0xb5, 0x2f, 0xfd}
+
+// checkCanCompress checks if there is an proposal payload message and peers supporting compression
+func checkCanCompress(request broadcastRequest, prio bool, peers []*wsPeer) bool {
+	canCompress := false
+	if prio {
+		hasPP := false
+		for _, tag := range request.tags {
+			if tag == protocol.ProposalPayloadTag {
+				hasPP = true
+				break
+			}
+		}
+		// if have proposal payload check if there are any peers supporting compression
+		if hasPP {
+			for _, peer := range peers {
+				if peer.vfCompressedProposalSupported() {
+					canCompress = true
+					break
+				}
+			}
+		}
+	}
+	return canCompress
+}
+
+// zstdCompressMsg returns a concatenation of a tag and compressed data
+func zstdCompressMsg(tbytes []byte, d []byte) ([]byte, string) {
+	bound := zstd.CompressBound(len(d))
+	mbytesComp := make([]byte, len(tbytes)+bound)
+	copy(mbytesComp, tbytes)
+	comp, err := zstd.Compress(mbytesComp[len(tbytes):], d)
+	if err != nil {
+		// fallback and reuse non-compressed original data
+		logMsg := fmt.Sprintf("failed to compress into buffer of len %d: %v", len(d), err)
+		copied := copy(mbytesComp[len(tbytes):], d)
+		return mbytesComp[:len(tbytes)+copied], logMsg
+	}
+	mbytesComp = mbytesComp[:len(tbytes)+len(comp)]
+	return mbytesComp, ""
+}
+
+// MaxDecompressedMessageSize defines a maximum decompressed data size
+// to prevent zip bombs
+const MaxDecompressedMessageSize = 20 * 1024 * 1024 // some large enough value
+
+// wsPeerMsgDataConverter performs optional incoming messages conversion.
+// At the moment it only supports zstd decompression for payload proposal
+type wsPeerMsgDataConverter struct {
+	log    logging.Logger
+	origin string
+
+	// actual converter(s)
+	ppdec zstdProposalDecompressor
+}
+
+type zstdProposalDecompressor struct {
+	active bool
+}
+
+func (dec zstdProposalDecompressor) enabled() bool {
+	return dec.active
+}
+
+func (dec zstdProposalDecompressor) accept(data []byte) bool {
+	return len(data) > 4 && bytes.Equal(data[:4], zstdCompressionMagic[:])
+}
```

Comment:
What would otherwise be at the front of messages?  Is it anything a user can control?

Reply:
raw proposal message: `unauthenticatedProposal` that starts with `Block` that starts with Round. At round 0xFD2FB528 = 4,247,762,216 in 267 years (assuming 2s round) this code would try to decompress non-compressed data.

Reply:
Are proposals not msgpacked?  Wouldn't there be a key, and the value then varuint encoded?  That would make me happier - if the msgpack key will come first, it's just plain impossible to mix up.

Reply:
I just mean that since I now know that the message is msgpacked, I feel much more confident that a proposal message will not mistakenly have the right magic at the front to confuse this code.